src/dove/core/transfers.py

```diff
@@ -92,33 +109,33 @@ def __call__(self, inputs: dict[str, float], outputs: dict[str, float]) -> pyo.E
 
         Returns
         -------
-        pyo.Expression
-            A constraint enforcing the ratio relation, or Constraint.Skip for Source/Sink cases.
+        list[float]
+            A list of values that must be equal for the transfer constraint to be satisfied.
 
         Raises
         ------
         ValueError
-            If neither input nor output resource is found in the provided dictionaries.
+            If an input or output resource cannot be found in the dispatch variables
         """
-        has_input = self.input_res.name in inputs
-        has_output = self.output_res.name in outputs
-
-        if has_input and has_output:
-            return outputs[self.output_res.name] == self.ratio * inputs[self.input_res.name]
-
-        elif has_output and not has_input:
-            # Source: output = output (tautology)
-            return pyo.Constraint.Skip
-
-        elif has_input and not has_output:
-            # Sink: input = input (tautology)
-            return pyo.Constraint.Skip
-
-        else:
-            raise ValueError(
-                f"RatioTransfer could not find either input '{self.input_res}' "
-                f"or output '{self.output_res}' in the provided dispatch variables."
-            )
+        for input_res in self.input_resources:
+            if input_res.name not in inputs:
+                raise ValueError(
+                    f"RatioTransfer: Input resource '{input_res.name}' "
+                    "not found in the provided dispatch variable for inputs"
+                )
+
+        for output_res in self.output_resources:
+            if output_res.name not in outputs:
+                raise ValueError(
+                    f"RatioTransfer: Output resource '{output_res.name}' "
+                    "not found in the provided dispatch variable for outputs"
+                )
+
+        weighted_inputs = [inputs[res.name] / ratio for res, ratio in self.input_resources.items()]
+        weighted_outputs = [
+            outputs[res.name] / ratio for res, ratio in self.output_resources.items()
+        ]
+        return weighted_inputs + weighted_outputs
```

Comment:
are signs considered here (i.e., inputs are + and outputs are -)? or somewhere downstream of this?

Reply:
perfect, thanks for the explanation! 

Context for my question:
I was confused because HERON dispatch handled things a little different but these are equivalent methods. In HERON, the components still tracked which resources were consumed and produced but it attached signs directly to the flow (there, the "production") variables. so the resource balance is a straight sum over all production variables (IO signs already incorporated). and the resource transfer constraints have negative signs in front of the outputs. 

I see now that the DOVE component flow variables are all inherently positive, the actual signs are implemented in the resource balance (all inputs get a + and all outputs get a -) summation.